File: modelers/matlab/core/RealJointDomain.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   Copyright 2012 Analog Devices, Inc.
%
%   Licensed under the Apache License, Version 2.0 (the "License");
%   you may not use this file except in compliance with the License.
%   You may obtain a copy of the License at
%
%       http://www.apache.org/licenses/LICENSE-2.0
%
%   Unless required by applicable law or agreed to in writing, software
%   distributed under the License is distributed on an "AS IS" BASIS,
%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%   See the License for the specific language governing permissions and
%   limitations under the License.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef RealJointDomain < Domain
    
    properties
        NumElements;
        RealDomains;
    end
    
    methods
        function obj = RealJointDomain(varargin)
            
            % Either a number or a vector of RealDomains
            if isempty(varargin)
                error('Constructor is missing arguments');
            elseif length(varargin) == 1
                arg = varargin{1};
                if isa(arg, 'com.analog.lyric.dimple.matlabproxy.PRealJointDomain')
                    obj.IDomain = arg;
                    obj.RealDomains = wrapProxyObject(arg.getRealDomains());
                    return;
                elseif isnumeric(arg)
                    domains = cell(arg,1);
                    for i = 1:length(domains)
                        domains{i} = RealDomain();
                    end
                else
                    domains = arg;
                end
            else
                if isnumeric(varargin{1})
                    dimension = varargin{1};
                    domains = varargin(2:end);
                    if (numel(domains) == 1)
                        domains = cell(1,dimension);
                        for i=1:dimension
                            domains{i} = varargin{2};
                        end
                    elseif (numel(domains) ~= dimension);
                        error('Number of domain elements must match the number of joint variable elements');
                    end
                else
                    domains = varargin;
                end
            end
            
            domains = domains(:);
            idomains = cell(size(domains));
            for i = 1:length(idomains)
                idomains{i} = domains{i}.IDomain;
            end
            
            modeler = getModeler();
            obj.IDomain = modeler.createRealJointDomain(idomains);
            obj.RealDomains = domains;
        end
        
        function retval = isDiscrete(obj)
            retval = false;
        end
        
        function numElements = get.NumElements(obj)
            numElements = obj.IDomain.getNumVars();
        end
    end
end

File: solvers/java/src/main/java/com/analog/lyric/dimple/model/RealJointDomain.java
/*******************************************************************************
*   Copyright 2012 Analog Devices, Inc.
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
********************************************************************************/

package com.analog.lyric.dimple.model;

import java.lang.reflect.Array;
import java.util.Arrays;

import net.jcip.annotations.Immutable;

@Immutable
public class RealJointDomain extends Domain
{
	/*-------
	 * State
	 */
	
	private static final long serialVersionUID = 1L;
	
	private final RealDomain [] _domains;
	
	/*--------------
	 * Construction
	 */
	
	RealJointDomain(RealDomain[] domains)
	{
		super(Arrays.hashCode(domains) * 3 + 31);
		_domains = domains;
	}

	public static RealJointDomain create(RealDomain... domains)
	{
		if (domains.length < 2)
		{
			throw new IllegalArgumentException("RealJointDomain requires at least two domains");
		}
		return new RealJointDomain(domains.clone());
	}

	/*----------------
	 * Object methods
	 */
	
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if ((obj instanceof RealJointDomain))
		{
			RealJointDomain other = (RealJointDomain) obj;
			return Arrays.equals(this._domains, other._domains);
		}
		
		return false;
	}
	
	/*----------------
	 * Domain methods
	 */

	@Override
	public final RealJointDomain asRealJoint()
	{
		return this;
	}
	
	/**
	 * @returns true if value is an array of length matching {@link #getNumVars()} and whose
	 * elements are contained in the correspond {@link RealDomain} in {@link #getRealDomains()}.
	 */
	@Override
	public boolean inDomain(Object value)
	{
		if (value.getClass().isArray() && Array.getLength(value) == _domains.length)
		{
			for (int i = 0, end = _domains.length; i < end; ++i)
			{
				if (! _domains[i].inDomain(Array.get(value, i)))
				{
					return false;
				}
			}
			
			return true;
		}
		
		return false;
	}
	
	public boolean inDomain(double ... value)
	{
		int size = value.length;
		if (size != _domains.length)
			return false;
		
		for (int i = 0; i < size; i++)
			if (!_domains[i].inDomain(value[i]))
				return false;
		
		return true;
	}

	@Override
	public final boolean isRealJoint()
	{
		return true;
	}
	
	/*-------------------------
	 * RealJointDomain methods
	 */
	
	/**
	 * Returns the number of dimensions, or subdomains that make up this joint domain.
	 */
	public final int getDimensions()
	{
		return _domains.length;
	}
	
	public RealDomain [] getRealDomains()
	{
		return _domains.clone();
	}
	
	public RealDomain getRealDomain(int dimension)
	{
		return _domains[dimension];
	}
	
	public int getNumVars()
	{
		return _domains.length;
	}
}
